### PR TITLE
Lesson 5

### PR DESCRIPTION
**Pull request by OmegaLord**

### CODE REVIEW

File: lesson-3/train.rb [hunk 1]
```diff
@@ -23,7 +37,7 @@ def hook_carriage(carriage)
   end
 
   def unhook_carriage
-    raise if speed.positive? || carriages.empty?
+    raise "Speed is'n zero or no carriages" if speed.positive? || carriages.empty?
```
Comment (AZatsepa) [inline lesson-3/train.rb]: You can raise a special message for each case

File: spec/cargo_train_spec.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  describe '#type' do
+    it { expect(cargo_train.type).to eq(type) }
+  end
+
+  describe '#hook_carriage' do
+    it { expect { cargo_train.hook_carriage(cargo_carriage) }.to change { cargo_train.carriages.length }.by(1) }
+  end
+
+  describe '#unhook_carriage' do # use "shared examples"
```
Comment (AZatsepa) [inline spec/cargo_train_spec.rb]: Is it possible to avoid 'describe' here?

File: spec/instance_counter_spec.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      include InstanceCounter
+      def initialize
+        register_instance
+      end
+    end
+  end
+  let(:my_instance) { my_class.new }
+
+  describe '#register_instance' do
+    it { expect { my_class.new }.to change(my_class, :instances).by(1) }
```
Comment (AZatsepa) [inline spec/instance_counter_spec.rb]: :+1:

File: spec/instance_counter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+describe InstanceCounter do
+  let(:my_class) do
+    Class.new do
+      include InstanceCounter
+      def initialize
+        register_instance
+      end
+    end
+  end
+  let(:my_instance) { my_class.new }
```
Comment (AZatsepa) [inline spec/instance_counter_spec.rb]: Where do you use it?

File: spec/manufacturer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+describe Manufacturer do
+  let(:my_class) { Class.new { include Manufacturer } }
+  let(:my_module) { my_class.new } # Is it a class, or a module?
```
Comment (AZatsepa) [inline spec/manufacturer_spec.rb]: A comment looks strange

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+require_relative '../lesson-3/route'
+require_relative '../lesson-3/carriage'
+require_relative '../lesson-3/cargo_train'
+require_relative '../lesson-3/cargo_carriage'
+require_relative '../lesson-3/passenger_train'
+require_relative '../lesson-3/passenger_carriage'
+require_relative '../lesson-3/manufacturer'
+require_relative '../lesson-3/instance_counter'
+# require_relative './train_unhook_carriage_spec'
+# Dir["./spec/../lesson-3/*.rb"].sort.each { |f| require f }
```
Comment (AZatsepa) [inline spec/spec_helper.rb]: Remove unused comments

File: spec/station_spec.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+describe Station do
+  let(:station) { described_class.new('station1') }
+  let(:station_2) { described_class.new('station2') }
+  let(:train) { Train.new(number) }
+  let(:number) { 'number' }
+
+  describe '#receive_train' do
+    it { expect(station.receive_train(train)).to include(train) }
```
Comment (AZatsepa) [inline spec/station_spec.rb]: You should test what method does, not what it returns

File: spec/station_spec.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  describe '#send_train' do
+    before do
+      station.receive_train(train)
+    end
+
+    it 'pops up train' do
+      expect(station.send_train(number)).to be(train)
```
Comment (AZatsepa) [inline spec/station_spec.rb]: You should test what method does, not what it returns

File: spec/support/train_unhook_carriage.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    train.hook_carriage(carriage)
+  end
+
+  it { expect { train.unhook_carriage }.to change { train.carriages.length }.by(-1) }
+
+  context 'when carriages number is zero' do
+    it { expect { 2.times { train.unhook_carriage } }.to raise_exception error_message }
+  end
+
+  context 'when train speed is not zero' do
```
Comment (AZatsepa) [inline spec/support/train_unhook_carriage.rb]:
```
context 'when speed is not zero' do
  it 'raises error' do
    train.accelerating_by(100)
   expect { train.unhook_carriage }.to raise_exception error_message
  end
end
# you don`t neet add 'before' if you have one expectation

File: lesson-3/station.rb [hunk 10]
```diff
@@ -3,20 +3,30 @@
...
+    trains.delete_at(trains.index { |t| t.number == train_number })
   end
 
   def trains_count_by_type(type)
     trains.count { |t| t.type == type }
   end
+
+  class << self
+    attr_reader :my_instances
+    def all
```
[the remark this answers is not in the document]
Comment (OmegaLord) [reply lesson-3/station.rb]: ok

**Pull request merged**